### ISSUE TITLE
New here, my bot always crashed in less than 30 minutes running, please help me ^^

### ISSUE THREAD

**Issue by Howning** [tweepy/tweepy#1205]
This is my code : 

import tweepy
from time import sleep
import random

consumer_key = ''
consumer_secret = ''
access_token = ''
access_token_secret = ''
auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_token_secret)
api = tweepy.API(auth)
i=0
while i < 10000:
    search1 = "follow everyone who likes"
    numberOfTweets1 = int("197")

    for tweet in tweepy.Cursor(api.search, search1).items(numberOfTweets1):
        try:
            sleep(random.randint(20, 24))
            tweet.favorite()
            print('Fav the tweet 1')
        except tweepy.TweepError as e:
                print(e.reason)
        except StopIteration:
            break
    sleep(random.randint(1000, 1444))
    print("Part 1 done, sleeping for a few minutes")

    search2 = "ifb"
    numberOfTweets2 = int("144")
    for tweet in tweepy.Cursor(api.search, search2).items(numberOfTweets2):
        try:
            sleep(random.randint(20, 24))
            tweet.favorite()
            print('Fav the tweet 2')
        except tweepy.TweepError as e:
                print(e.reason)
        except StopIteration:
            break
    sleep(random.randint(1000, 1444))
    print("Part 2 done, sleeping for a few minutes")

    search3 = "sdv"
    numberOfTweets3 = int("158")
    for tweet in tweepy.Cursor(api.search, search3).items(numberOfTweets3):
        try:
            sleep(random.randint(20, 24))
            tweet.favorite()
            print('Fav the tweet 3')
        except tweepy.TweepError as e:
                print(e.reason)
        except StopIteration:
            break
    sleep(random.randint(1000, 1444))
    print("Part 3 done, sleeping for a few minutes")

    search4 = "Siga todos que FAV"
    numberOfTweets4 = int("117")
    for tweet in tweepy.Cursor(api.search, search4).items(numberOfTweets4):
        try:
            sleep(random.randint(20, 24))
            tweet.favorite()
            print('Fav the tweet 4')
        except tweepy.TweepError as e:
                print(e.reason)
        except StopIteration:
            break

    sleep(random.randint(1000,1444))
    print("Part 4 done, sleeping for a few minutes")

    sleep(random.randint(7200, 10000))
    i += 1

And that's the error I always get in less than 30 minutes when I start run my bot : 
Traceback (most recent call last):

  File "C:\Users\Utilisateur\Desktop\Dev\Python\Lib\site-packages\urllib3\connectionpool.py", line 600, in urlopen
    chunked=chunked)
  File "C:\Users\Utilisateur\Desktop\Dev\Python\Lib\site-packages\urllib3\connectionpool.py", line 384, in _make_request
    six.raise_from(e, None)
  File "<string>", line 2, in raise_from
  File "C:\Users\Utilisateur\Desktop\Dev\Python\Lib\site-packages\urllib3\connectionpool.py", line 380, in _make_request
    httplib_response = conn.getresponse()
  File "C:\Users\Utilisateur\Desktop\Dev\Python\lib\http\client.py", line 1321, in getresponse
    response.begin()
  File "C:\Users\Utilisateur\Desktop\Dev\Python\lib\http\client.py", line 296, in begin
    version, status, reason = self._read_status()
  File "C:\Users\Utilisateur\Desktop\Dev\Python\lib\http\client.py", line 257, in _read_status
    line = str(self.fp.readline(_MAXLINE + 1), "iso-8859-1")
  File "C:\Users\Utilisateur\Desktop\Dev\Python\lib\socket.py", line 589, in readinto
    return self._sock.recv_into(b)
  File "C:\Users\Utilisateur\Desktop\Dev\Python\lib\ssl.py", line 1052, in recv_into
    return self.read(nbytes, buffer)
  File "C:\Users\Utilisateur\Desktop\Dev\Python\lib\ssl.py", line 911, in read
    return self._sslobj.read(len, buffer)
ConnectionResetError: [WinError 10054] Une connexion existante a dû être fermée par l’hôte distant

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\Utilisateur\Desktop\Dev\Python\Lib\site-packages\requests\adapters.py", line 449, in send
    timeout=timeout
  File "C:\Users\Utilisateur\Desktop\Dev\Python\Lib\site-packages\urllib3\connectionpool.py", line 638, in urlopen
    _stacktrace=sys.exc_info()[2])
  File "C:\Users\Utilisateur\Desktop\Dev\Python\Lib\site-packages\urllib3\util\retry.py", line 368, in increment
    raise six.reraise(type(error), error, _stacktrace)
  File "C:\Users\Utilisateur\Desktop\Dev\Python\Lib\site-packages\urllib3\packages\six.py", line 685, in reraise
    raise value.with_traceback(tb)
  File "C:\Users\Utilisateur\Desktop\Dev\Python\Lib\site-packages\urllib3\connectionpool.py", line 600, in urlopen
    chunked=chunked)
  File "C:\Users\Utilisateur\Desktop\Dev\Python\Lib\site-packages\urllib3\connectionpool.py", line 384, in _make_request
    six.raise_from(e, None)
  File "<string>", line 2, in raise_from
  File "C:\Users\Utilisateur\Desktop\Dev\Python\Lib\site-packages\urllib3\connectionpool.py", line 380, in _make_request
    httplib_response = conn.getresponse()
  File "C:\Users\Utilisateur\Desktop\Dev\Python\lib\http\client.py", line 1321, in getresponse
    response.begin()
  File "C:\Users\Utilisateur\Desktop\Dev\Python\lib\http\client.py", line 296, in begin
    version, status, reason = self._read_status()
  File "C:\Users\Utilisateur\Desktop\Dev\Python\lib\http\client.py", line 257, in _read_status
    line = str(self.fp.readline(_MAXLINE + 1), "iso-8859-1")
  File "C:\Users\Utilisateur\Desktop\Dev\Python\lib\socket.py", line 589, in readinto
    return self._sock.recv_into(b)
  File "C:\Users\Utilisateur\Desktop\Dev\Python\lib\ssl.py", line 1052, in recv_into
    return self.read(nbytes, buffer)
  File "C:\Users\Utilisateur\Desktop\Dev\Python\lib\ssl.py", line 911, in read
    return self._sslobj.read(len, buffer)
urllib3.exceptions.ProtocolError: ('Connection aborted.', ConnectionResetError(10054, 'Une connexion existante a dû être fermée par l’hôte distant', None, 10054, None))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\Utilisateur\Desktop\Dev\Python\Lib\site-packages\tweepy\binder.py", line 190, in execute
    proxies=self.api.proxy)
  File "C:\Users\Utilisateur\Desktop\Dev\Python\Lib\site-packages\requests\sessions.py", line 533, in request
    resp = self.send(prep, **send_kwargs)
  File "C:\Users\Utilisateur\Desktop\Dev\Python\Lib\site-packages\requests\sessions.py", line 646, in send
    r = adapter.send(request, **kwargs)
  File "C:\Users\Utilisateur\Desktop\Dev\Python\Lib\site-packages\requests\adapters.py", line 498, in send
    raise ConnectionError(err, request=request)
requests.exceptions.ConnectionError: ('Connection aborted.', ConnectionResetError(10054, 'Une connexion existante a dû être fermée par l’hôte distant', None, 10054, None))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:/Users/Utilisateur/Desktop/Dev/Python/Lib/site-packages/ThomasTwitter.py", line 18, in <module>
    for tweet in tweepy.Cursor(api.search, search1).items(numberOfTweets1):
  File "C:\Users\Utilisateur\Desktop\Dev\Python\Lib\site-packages\tweepy\cursor.py", line 49, in __next__
    return self.next()
  File "C:\Users\Utilisateur\Desktop\Dev\Python\Lib\site-packages\tweepy\cursor.py", line 197, in next
    self.current_page = self.page_iterator.next()
  File "C:\Users\Utilisateur\Desktop\Dev\Python\Lib\site-packages\tweepy\cursor.py", line 108, in next
    data = self.method(max_id=self.max_id, parser=RawParser(), *self.args, **self.kargs)
  File "C:\Users\Utilisateur\Desktop\Dev\Python\Lib\site-packages\tweepy\binder.py", line 250, in _call
    return method.execute()
  File "C:\Users\Utilisateur\Desktop\Dev\Python\Lib\site-packages\tweepy\binder.py", line 192, in execute
    six.reraise(TweepError, TweepError('Failed to send request: %s' % e), sys.exc_info()[2])
  File "C:\Users\Utilisateur\Desktop\Dev\Python\Lib\site-packages\six.py", line 692, in reraise
    raise value.with_traceback(tb)
  File "C:\Users\Utilisateur\Desktop\Dev\Python\Lib\site-packages\tweepy\binder.py", line 190, in execute
    proxies=self.api.proxy)
  File "C:\Users\Utilisateur\Desktop\Dev\Python\Lib\site-packages\requests\sessions.py", line 533, in request
    resp = self.send(prep, **send_kwargs)
  File "C:\Users\Utilisateur\Desktop\Dev\Python\Lib\site-packages\requests\sessions.py", line 646, in send
    r = adapter.send(request, **kwargs)
  File "C:\Users\Utilisateur\Desktop\Dev\Python\Lib\site-packages\requests\adapters.py", line 498, in send
    raise ConnectionError(err, request=request)
tweepy.error.TweepError: Failed to send request: ('Connection aborted.', ConnectionResetError(10054, 'Une connexion existante a dû être fermée par l’hôte distant', None, 10054, None))

I have already tried a few solutions proposed on the forum but I can't fix that problem, it would be very kind if one of you can help me to solve that problem (I'm new here, ask me if you have questions and if you want more informations to help me solve my issue), thank you so much guys for your future help ^^

**Comment by machine-uprising** [comment]
I could consistently reproduce this error. Haven't nailed down the exact cause but it raises a requests.ConnectionError exception - Errno 10054 An existing connection was forcibly closed by the remote host - so there is some problem with the request being accepted by the server.
Capturing the request being sent when the connection error occurs and testing it manually shows the request is valid, it's just the server is not accepting the request. This can be for a few reasons (eg, server is unavailable, connection is already open and hasn't been closed before new connection is received, etc). Interestingly the connection error did not occur if I removed the sleep timers but rate limits then get triggered pretty quickly.

It's possible to add a fix to the execute() function in binder.py to handle these connection errors and keep the process running. Catching the ConnectionError exception and forcing a retry of the request resolved the problem in my tests. I also found this useful if I dropped and reconnected my network connection while running the code so the fix would handled multiple connection error issues. It involves adding a timer, a connection_timeout value, and wrapping the self.session.request(...) block in a while loop so that if connection errors occur there can be a number of retry attempts before concluding that the connection is unavailable and raising a TweepError. 

Bug Fix for binder.py
```python
#binder.py
def execute(self):
    ...
    ...
    # Execute request
    connection_timeout = 30
    start_time = time.time()
    while True:
        try:
            resp = self.session.request(self.method,
                                        full_url,
                                        data=self.post_data,
                                        timeout=self.api.timeout,
                                        auth=auth,
                                        proxies=self.api.proxy)
            break
        except requests.ConnectionError as ce:        
            if time.time() > start_time + connection_timeout:
                raise TweepError('ConnectionError: Could not connect within %s seconds' % connection_timeout)
            else:
                time.sleep(1)
        except Exception as e:
            six.reraise(TweepError, TweepError('Failed to send request: %s' % e), sys.exc_info()[2])
```

Another option for the connection_timeout variable is to move it into the API object (api.py) and pick it up in bind_api() similar to the retry_count, retry_errors, timeout variables. This way the connection_timeout value could be configured by users and not be a fixed value.

**Comment by machine-uprising** [comment]
@howning - as I wrote in the comment above, the error you receive is happening in Tweepy but the way you have written your code seems like it would cause you problems as well. 
With all the sleep intervals you have this code could run for a minimum of 2722 days. Each iteration of the searches inside your while loop can run a minimum of 6.53 hours. You're then iterating this 10000 times (65300 hours). Another problem is the while loop iteration is going to duplicate tweets because you don't have anything to check if the tweet in search results has already been captured in a previous iteration. I ran a test where after several iterations there were about 9000 tweets pulled from the searches but there were only about 700 unique tweets. You'll be wasting a lot of API requests if you're trying to favourite the same tweets multiple times. 

One thing you can do is increase the number of search results you receive per API call. The search endpoint has a 'count' variable to specify how many search results to retrieve per API call. By default this is only 15 but you can set the 'count' variable up to 100. This will reduce the number of searches you're running. 

```python 
for tweet in tweepy.Cursor(api.search, search_string, count=100).items(numberOfTweets):
```
Maybe what you could do is capture a list of tweet IDs from the search and then run a separate iteration to favourite those tweets. For example:
```python
tweet_id_list = []
for tweet in tweepy.Cursor(api.search, search_string, count=count).items(limit):
    try:
        tweet_id_list.append(tweet._json['id'])
    except tweepy.TweepError as e:
        print(e.reason)
    except StopIteration:
        break

# now favourite the tweets in tweet_id_list
# use a set to remove duplicate tweet_id from the list 
for tweet_id in set(tweet_id_list):
    api.create_favorite(tweet_id)
```
With the volume of tweets your code is dealing with you will more than likely hit rate limits so you would need ways to handle these which I think is why you put all the sleep timers. Tweepy does have built-in rate limit handling. Maybe you write the tweet id list to a file and run the create_favorite() as a separate python process. 

Except for the connection error these are not really issues with tweepy so you will need to search tweepy documentation, stackoverflow or other resources for more info on how to refactor your code.

**Comment by johnkw** [comment]
The above patch would be helpful in some cases, but really this needs to be handled with a user-code callback function, so the user code can determine how long to sleep, or whether to abort, etc.  I have a longer list of transient errors which I'm currently retrying on. It includes SSL stuff, but also twitter.com stuff like the 'Over capacity' error, which is API equivalent of the fail whale.

If this project is becoming active again I'll add a pull request for that. For now I'll wait on PR #1211 getting incorporated.

**Comment by johnkw** [comment]
I filed a clean issue #1219 for this. This ticket should probably be closed in favor of that one.

**Comment by Harmon758** [comment]
Closing this in favor of #1606
